File: blind75/67_implement_trie/trie.py
class TrieNode:
    """Represents a node in the trie."""

    def __init__(self):
        self.children = {}
        self.is_end_of_word = False


class Trie:
    """A prefix tree (trie) data structure for efficient string operations."""

    def __init__(self):
        """Initializes the trie object."""
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        Time complexity: O(m) where m is the length of the word.

        Args:
            word: The word to insert
        """
        current = self.root

        for ch in word:
            if ch not in current.children:
                current.children[ch] = TrieNode()
            current = current.children[ch]

        current.is_end_of_word = True

    def search(self, word: str) -> bool:
        """
        Returns true if the word is in the trie.
        Time complexity: O(m) where m is the length of the word.

        Args:
            word: The word to search for

        Returns:
            True if the word exists in the trie, False otherwise
        """
        current = self.root

        for ch in word:
            if ch not in current.children:
                return False
            current = current.children[ch]

        return current.is_end_of_word

    def starts_with(self, prefix: str) -> bool:
        """
        Returns true if there is any word in the trie that starts with the given prefix.
        Time complexity: O(m) where m is the length of the prefix.

        Args:
            prefix: The prefix to search for

        Returns:
            True if any word has this prefix, False otherwise
        """
        current = self.root

        for ch in prefix:
            if ch not in current.children:
                return False
            current = current.children[ch]

        return True

File: blind75/67_implement_trie/trie.go
package trie

// TrieNode represents a node in the trie.
type TrieNode struct {
	children    map[rune]*TrieNode
	isEndOfWord bool
}

// NewTrieNode creates a new trie node.
func NewTrieNode() *TrieNode {
	return &TrieNode{
		children:    make(map[rune]*TrieNode),
		isEndOfWord: false,
	}
}

// Trie represents a prefix tree data structure.
type Trie struct {
	root *TrieNode
}

// Constructor initializes the trie object.
func Constructor() Trie {
	return Trie{
		root: NewTrieNode(),
	}
}

// Insert inserts a word into the trie.
// Time complexity: O(m) where m is the length of the word.
func (t *Trie) Insert(word string) {
	current := t.root

	for _, ch := range word {
		if _, exists := current.children[ch]; !exists {
			current.children[ch] = NewTrieNode()
		}
		current = current.children[ch]
	}

	current.isEndOfWord = true
}

// Search returns true if the word is in the trie.
// Time complexity: O(m) where m is the length of the word.
func (t *Trie) Search(word string) bool {
	current := t.root

	for _, ch := range word {
		if _, exists := current.children[ch]; !exists {
			return false
		}
		current = current.children[ch]
	}

	return current.isEndOfWord
}

// StartsWith returns true if there is any word in the trie that starts with the given prefix.
// Time complexity: O(m) where m is the length of the prefix.
func (t *Trie) StartsWith(prefix string) bool {
	current := t.root

	for _, ch := range prefix {
		if _, exists := current.children[ch]; !exists {
			return false
		}
		current = current.children[ch]
	}

	return true
}
